### PR TITLE
Added turn logging.

### PR DESCRIPTION
**Pull request by oskopek**

### CODE REVIEW

File: codenames/supervised/agents.py [hunk 1]
```diff
@@ -22,6 +22,7 @@ def train_step(self, encoder_hidden, encoder_outputs, clue_logits, intended_outp
         clue_scaled = self.clue_scaler(clue_logits)
```
Comment (VIGS25) [inline codenames/supervised/agents.py]: `PlayerTrainer` is guesser. The guesser shouldn't have access to all the `clue_logits` since the `SpyMaster` gives out only one clue. We need to fix this.

File: codenames/supervised/train.py [hunk 2]
```diff
@@ -120,7 +120,23 @@ def _save_model(self):
...
 
-    def _train_step(self, board_input, criterion, team_word_mask, att_max_len=data.BOARD_TENSOR_MAX_LENGTH):
+    def _print_turn(self, team, board_input, clue_logits, intended_output_idx, guessed_card_idx):
+        board = [(self.board_vocab.idx_to_word[idx], team) for idx, team in board_input]
+        board_words = [word for word, _ in board]
+
+        clue_word_idx = torch.argmax(clue_logits, dim=-1)
+        clue_word = self.clue_vocab.idx_to_word[clue_word_idx]
+        intended_output_words = set(board_words[idx] for idx in intended_output_idx)
+        guessed_words = [board_words[idx] for idx in guessed_card_idx]
```
Comment (VIGS25) [inline codenames/supervised/train.py]: Do we want to print <EOS> as well?

File: codenames/supervised/util.py [hunk 3]
```diff
@@ -36,3 +36,18 @@ def cat_params(*args):
 
 def printf(*args):
     print(*args, flush=True)
+
+
+def print_board(board):
```
Comment (VIGS25) [inline codenames/supervised/util.py]: I think the guy who implemented only the spymaster had a nice printing thing for the board. He also updated guessed words to their corresponding colors or teams. Can we add that?

File: codenames/supervised/agents.py [hunk 4]
```diff
@@ -22,6 +22,7 @@ def train_step(self, encoder_hidden, encoder_outputs, clue_logits, intended_outp
...
         decoder_hidden = torch.cat([encoder_hidden, clue_scaled], -1)
         decoder_input = torch.tensor(self.board_vocab.word_to_idx[data.SOS_TOKEN], dtype=torch.long, device=device)
+        guessed_card_idxs = []
         for di in range(intended_output.size(0)):
```
Comment (VIGS25) [inline codenames/supervised/agents.py]: I think, should be `intended_output.size(0) + 1`, but we can add this later.

File: codenames/supervised/util.py (hunk 3 above)
Comment (oskopek) [reply codenames/supervised/util.py]: Sure, go crazy. I wanted this to work even in Leonhard text files, so I'm good with this, honestly.

File: codenames/supervised/agents.py (hunk 4 above)
Comment (oskopek) [reply codenames/supervised/agents.py]: Well, there is an EOS option, right? So that's effectively pass. And yeah, we can add + 1 here, wouldn't make much of a difference thought, IMO. Fixing.

File: codenames/supervised/train.py (hunk 2 above)
Comment (oskopek) [reply codenames/supervised/train.py]: I do for now, just for debugging purposes. The actual UI can be different later.

File: codenames/supervised/agents.py (hunk 4 above)
Comment (VIGS25) [reply codenames/supervised/agents.py]: Not a big one this, can be ignored :)

File: codenames/supervised/agents.py (hunk 1 above)
Comment (oskopek) [reply codenames/supervised/agents.py]: Renamed to `PlayerTrainer` to `Guesser`. Am thinking about how to properly pass the clue.

Comment (VIGS25) [reply codenames/supervised/agents.py]: Can we sample one clue based on logits and just pass that?

Comment (oskopek) [reply codenames/supervised/agents.py]: Fixed, please take a look :)

File: codenames/supervised/agents.py [hunk 5]
```diff
@@ -1,28 +1,45 @@
...
-        self.clue_scaler = nn.Linear(clues_vocab_size, enc_hidden_size)
-        self.decoder = network.AttnDecoderRNN(enc_hidden_size, dec_hidden_size, data.BOARD_SIZE)
-        self.models = [self.clue_scaler, self.decoder]
+        self.clue_vocab = clue_vocab
+        self.clue_embed = clue_embed
+        enc_hidden_size = self.clue_embed.embedding_dim
+        self.encoder = network.EncoderOutputRNN(hidden_size=enc_hidden_size, output_size=dec_hidden_size)
+        self.decoder = network.AttnDecoderRNN(dec_hidden_size, dec_hidden_size + self.clue_embed.embedding_dim, data.BOARD_SIZE)
+        self.models = [self.encoder, self.decoder]
+        self.device = "cpu"
```
Comment (VIGS25) [inline codenames/supervised/agents.py]: `device = "cpu"` on purpose?
